README.md
# [MetovaJSONCodable](https://cocoapods.org/pods/MetovaJSONCodable)

<p align="center">
<a href="https://travis-ci.org/metova/MetovaJSONCodable" target="_blank"><img src="https://travis-ci.org/metova/MetovaJSONCodable.svg?branch=master" alt="Build Status"></a> 
<a href="https://cocoapods.org/pods/MetovaJSONCodable" target="_blank"><img src="https://img.shields.io/cocoapods/v/MetovaJSONCodable.svg" alt="CocoaPods Compatible"/></a>
<a href="http://metova.github.io/MetovaJSONCodable/" target="_blank"><img src="https://cdn.rawgit.com/metova/MetovaJSONCodable/master/docs/badge.svg" alt="Documentation"/></a>
<a href="https://coveralls.io/github/metova/MetovaJSONCodable?branch=master" target="_blank"><img src="https://coveralls.io/repos/github/metova/MetovaJSONCodable/badge.svg?branch=master&dummy=no_cache_please_1" alt="Coverage Status"/></a>
<a href="http://cocoadocs.org/docsets/MetovaJSONCodable" target="_blank"><img src="https://img.shields.io/cocoapods/p/MetovaJSONCodable.svg?style=flat" alt="Platform"/></a>
<a href="http://twitter.com/metova" target="_blank"><img src="https://img.shields.io/badge/twitter-@Metova-3CAC84.svg" alt="Twitter"/></a>
<br/>
</p>

MetovaJSONCodable is a lightweight Codable protocol with addtional support for JSON

- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
    - [Object example](#object-example)
    - [JSON](#json)
    - [Custom Date Handling](#custom-date-handling)
- [Documentation](#documentation)
- [Credits](#credits)
- [License](#license)

## Requirements

- Swift 4.2
- iOS 9+

## Installation

MetovaJSONCodable is available through [CocoaPods](http://cocoapods.org). To install it, simply add the following line to your Podfile:

```ruby
pod 'MetovaJSONCodable'
```

## Usage

Any object that previously conformed to the Codable protocol can now use the JSONCodable protocol.

#### Object example

```swift
struct Token: JSONCodable {

   var tokenID: Int
   var email: String
   var authenticationToken: String
}
```

#### JSON

```swift
let tokenJSON: JSON = [
    "testID": 1,
    "email": "test@email.com",
    "authenticationToken": "token"
]
    
let token = Token(from: tokenJSON)
```

#### Custom Date Handling

This pod comes with several child protocols for handling json with custom date formats

#### Time interval since reference date

```
struct Token: JSONEpochDateCodable {
    var tokenID: Int
    var email: String
    var authenticationToken: String
    var expirationDate: Date
}

let tokenJSON: JSON = [
    "tokenID": 1,
    "email": "test@email.com",
    "authenticationToken": "token",
    "expirationDate": 1523560533
]
    
let token = Token(from: tokenJSON)
    
// token.expirationDate is a Date equaling 2018-04-12 19:15:33 +0000
```

#### Custom Date foramatter object

We also have a child protocol for supporting a custom date formatter. Here you will need to implement a dateFormatter computed variable that returns a date formatter in order to conform to the protocol

```
struct Token: JSONCustomDateFormatterCodable {
    var testID: Int
    var email: String
    var expirationDate: Date

    static var dateFormatter: DateFormatter {

        let dateFormatter = DateFormatter()
        dateFormatter.calendar = Calendar(identifier: .iso8601)
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
        return dateFormatter
    }
}

let validTestJSONISO8601: JSON = [
    "testID": 2,
    "email": "test@email.com",
    "expirationDate": "2016-12-31T17:56:27.000Z"
]

let token = Token(from: tokenJSON)

// token.expirationDate is a Date equaling 2016-12-31 17:56:27 +0000
```

#### Custom Child Protocols

If you want to define your own date formatting strategy or some other permutation to the default coder/decoder, you simply need to make a child protocol and implement your own ecoder/coder

```
public protocol JSONDecodeableIOS8601: JSONDecodable {}

public extension JSONDecodeableIOS8601 {

    static var jsonDecoder: JSONDecoder {

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }
}
```


## Documentation

Documentation can be found [here](http://metova.github.io/MetovaJSONCodable/).

## Credits

MetovaJSONCodable is owned and maintained by [Metova Inc.](https://metova.com)

[Contributors](https://github.com/Metova/MetovaJSONCodable/graphs/contributors)

## License

MetovaJSONCodable is available under the MIT license. See the [LICENSE](LICENSE) file for more info.


CHANGELOG.md
# MetovaJSONCodable CHANGELOG

## 2.0.0

- Updated to support Swift 4.2

## 1.0.0

- Initial release.


MetovaJSONCodable/JSONCodable.swift
//
//  JSONCodable.swift
//  MetovaJSONCodable
//
//  Created by Kalan Stowe on 4/10/18.
//  Copyright © 2018 Metova Inc.
//
//  MIT License
//
//  Permission is hereby granted, free of charge, to any person obtaining
//  a copy of this software and associated documentation files (the
//  "Software"), to deal in the Software without restriction, including
//  without limitation the rights to use, copy, modify, merge, publish,
//  distribute, sublicense, and/or sell copies of the Software, and to
//  permit persons to whom the Software is furnished to do so, subject to
//  the following conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
//  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
//  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
//  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import Foundation

/**
 Type alias for conforming to JSONDecodable & JSONEncodable. The same way Codable is a type alias for Decodable & Encodable
 */
public typealias JSONCodable = JSONDecodable & JSONEncodable

/**
 Type alias added for use as a shorthand json dictionary objects
 */
public typealias JSON = [String: Any]

/**
 JSONDecodable protocol with an init for JSON. Conforms to Decodable protocol
 */
public protocol JSONDecodable: Decodable {
    
    /**
     init from JSON object
     */
    init?(from json: JSON)
    
    /**
     init from JSON data
     */
    init?(from data: Data)
    
    /**
     Default JSONDecoder. This is mainly to be used by child protocols that need custom decoders
     */
    static var jsonDecoder: JSONDecoder { get }
}

public extension JSONDecodable {
    
    /**
     -return basic JSONDecoder. This is meant to be used by child protocols that need custom decoders
     */
    static var jsonDecoder: JSONDecoder {
        
        return JSONDecoder()
    }
    
    /**
     - return JSONDecodable object from JSON or nil if the decode fails
     */
    init?(from json: JSON) {
        
        do {
            let data = try JSONSerialization.data(withJSONObject: json)
            self = try Self.jsonDecoder.decode(Self.self, from: data)
        }
        catch {
            return nil
        }
    }
    
    /**
     - return JSONDecodable object from JSON Data or nil if the decode fails
     */
    init?(from data: Data) {
        
        do {
            self = try JSONDecoder().decode(Self.self, from: data)
        }
        catch {
            return nil
        }
    }
}

/**
 JSONEncodable protocol for easily getting a JSON object from a conforming JSONCodable object. Conforms to Encodable protocol
 */
public protocol JSONEncodable: Encodable {
    
    /**
     JSON representation of JSONCodable object
     */
    var jsonValue: JSON? { get }
    
    /**
     JSON data representation of JSONCodable object
     */
    var jsonData: Data? { get }
    
    /**
    * Deafult JSONEncoder. This is mainly to be used by child protocols that need custom encoders
    */
    static var jsonEncoder: JSONEncoder { get }
}

public extension JSONEncodable {
    
    /**
     -return basic JSONEncoder. This is meant to be used by child protocols that need custom encoders
     */
    static var jsonEncoder: JSONEncoder {
        
        return JSONEncoder()
    }
    
    /**
     - return data represendation of the JSONCodable object or nil if the encode fails
     */
    var jsonData: Data? {
        return try? Self.jsonEncoder.encode(self)
    }
    
    /**
     - return JSON represendation of the JSONCodable object or nil if the encode fails
     */
    var jsonValue: JSON? {
        
        guard let jsonData = jsonData else { return nil }
        
        do {
            let json = try JSONSerialization.jsonObject(with: jsonData) as? JSON
            return json
        }
        catch {
            return nil
        }
    }
}


MetovaJSONCodable/MetovaJSONCodable.h
//
//  MetovaJSONCodable.h
//  MetovaJSONCodable
//
//  Created by Kalan Stowe on 4/10/18.
//  Copyright © 2018 Metova Inc.
//
//  MIT License
//
//  Permission is hereby granted, free of charge, to any person obtaining
//  a copy of this software and associated documentation files (the
//  "Software"), to deal in the Software without restriction, including
//  without limitation the rights to use, copy, modify, merge, publish,
//  distribute, sublicense, and/or sell copies of the Software, and to
//  permit persons to whom the Software is furnished to do so, subject to
//  the following conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
//  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
//  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
//  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

@import Foundation;

//! Project version number for MetovaJSONCodable.
FOUNDATION_EXPORT double MetovaJSONCodableVersionNumber;

//! Project version string for MetovaJSONCodable.
FOUNDATION_EXPORT const unsigned char MetovaJSONCodableVersionString[];




MetovaJSONCodable/JSONCodable+DateHandling.swift
//
//  JSONCodable+DateHandling.swift
//  MetovaJSONCodable
//
//  Created by Kalan Stowe on 4/12/18.
//  Copyright © 2018 Metova Inc.
//
//  MIT License
//
//  Permission is hereby granted, free of charge, to any person obtaining
//  a copy of this software and associated documentation files (the
//  "Software"), to deal in the Software without restriction, including
//  without limitation the rights to use, copy, modify, merge, publish,
//  distribute, sublicense, and/or sell copies of the Software, and to
//  permit persons to whom the Software is furnished to do so, subject to
//  the following conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
//  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
//  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
//  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import Foundation

// MARK: JSONEpochDateCodable

/**
 Type alias for conforming to JSONEpochDateDecodable & JSONEpochDateEncodable. The same way Codable is a type alias for Decodable & Encodable
 */
typealias JSONEpochDateCodable = JSONEpochDateDecodable & JSONEpochDateEncodable

/**
 JSONDecodable child protocol with support for dates as timestamps
 */
public protocol JSONEpochDateDecodable: JSONDecodable {}

public extension JSONEpochDateDecodable {
    
    /**
     - return JSONDecoder with date decoding strategy for Seconds since 1970
     */
    static var jsonDecoder: JSONDecoder {
        
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .secondsSince1970
        return decoder
    }
}

/**
 JSONEncodable child protocol with support for dates as timestamps
 */
public protocol JSONEpochDateEncodable: JSONEncodable {}

public extension JSONEpochDateEncodable {
    
    /**
     - return JSONDecoder with date encoding strategy for Seconds since 1970
     */
    static var jsonEncoder: JSONEncoder {
        
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .secondsSince1970
        return encoder
    }
}

// MARK: JSONCodableWithCustomFormatter

/**
 Type alias for conforming to JSONCustomDateFormatterDecodable & JSONCustomDateFormatterEncodable. The same way Codable is a type alias for Decodable & Encodable
 */
public typealias JSONCustomDateFormatterCodable = JSONCustomDateFormatterDecodable & JSONCustomDateFormatterEncodable

/**
 JSONDecodable child protocol that allows the user to define a cust dateFormatter. There is no default dateFormatter defined here so any object conforming to the protocol will need to add this
 */
public protocol JSONCustomDateFormatterDecodable: JSONDecodable {
    
    /**
     Customizeable date formatter. This will need to be implemented by conforming objects
     */
    static var dateFormatter: DateFormatter { get }
}

public extension JSONCustomDateFormatterDecodable {
    
    /**
     - return JSONDecoder with custom date formatter that is user definable
     */
    static var jsonDecoder: JSONDecoder {
        
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .formatted(Self.dateFormatter)
        return decoder
    }
}

/**
 JSONEncodable child protocol that allows the user to define a cust dateFormatter. There is no default dateFormatter defined here so any object conforming to the protocol will need to add this
 */
public protocol JSONCustomDateFormatterEncodable: JSONEncodable {
    
    /**
     Customizeable date formatter. This will need to be implemented by conforming objects
     */
    static var dateFormatter: DateFormatter { get }
}

public extension JSONCustomDateFormatterEncodable {
    
    /**
     - return JSONEncoder with custom date formatter that is user definable
     */
    static var jsonEncoder: JSONEncoder {
        
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .formatted(Self.dateFormatter)
        return encoder
    }
}


MetovaJSONCodableTests/MetovaJSONCodableTests.swift
//
//  MetovaJSONCodableTests.swift
//  MetovaJSONCodableTests
//
//  Created by Kalan Stowe on 4/10/18.
//  Copyright © 2018 Metova Inc.
//
//  MIT License
//
//  Permission is hereby granted, free of charge, to any person obtaining
//  a copy of this software and associated documentation files (the
//  "Software"), to deal in the Software without restriction, including
//  without limitation the rights to use, copy, modify, merge, publish,
//  distribute, sublicense, and/or sell copies of the Software, and to
//  permit persons to whom the Software is furnished to do so, subject to
//  the following conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
//  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
//  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
//  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import XCTest
@testable import MetovaJSONCodable

struct Test: JSONCodable {
    var testID: Int
    var email: String
    var testOptional: String?
    var testEnum: TestStringEnum
}

enum TestStringEnum: String, Codable {
    case testMatch
    case testNoMatch = "NoMatch"
}

class MetovaJSONCodableTests: XCTestCase {
    
    let validTestJSON: JSON = [
        "testID": 1,
        "email": "test@email.com",
        "testOptional": "optional",
        "testEnum": "testMatch"
    ]
    
    let validTestJSONWithNil: JSON = [
        "testID": 2,
        "email": "test@optional.com",
        "testOptional": Optional<String>.none as Any,
        "testEnum": "testMatch"
    ]
    
    let invalidTestJSON: JSON = [
        "testID": 3,
        "email": "test@email.com",
        "testEnum": "failState"
    ]
    
    func testDecodeSuccess() {
        
        guard let testMatch = Test(from: validTestJSON) else {
            XCTFail("Failed to decode Test object from json: \(validTestJSON)")
            return
        }
        
        XCTAssertEqual(testMatch.testID, 1)
        XCTAssertEqual(testMatch.email, "test@email.com")
        XCTAssertEqual(testMatch.testOptional, "optional")
        XCTAssertEqual(testMatch.testEnum, TestStringEnum.testMatch)
    }
    
    func testDecodeFailure() {
        
        let testNil = Test(from: invalidTestJSON)
        
        XCTAssertNil(testNil)
    }
    
    func testDecodeSuccessWithNil() {
        
        guard let testWithNil = Test(from: validTestJSONWithNil) else {
            XCTFail("Failed to decode Test object from json: \(validTestJSONWithNil)")
            return
        }
        
        XCTAssertEqual(testWithNil.testID, 2)
        XCTAssertEqual(testWithNil.email, "test@optional.com")
        XCTAssertNil(testWithNil.testOptional)
        XCTAssertEqual(testWithNil.testEnum, TestStringEnum.testMatch)
    }
    
    func testDecodeFromDataSuccess() {
        
        do {
            let data = try JSONSerialization.data(withJSONObject: validTestJSON)
            
            guard let testFromData = Test(from: data) else {
                XCTFail("Failed to decode Test object from data")
                return
            }
            
            XCTAssertEqual(testFromData.testID, 1)
            XCTAssertEqual(testFromData.email, "test@email.com")
            XCTAssertEqual(testFromData.testEnum, TestStringEnum.testMatch)
        }
        catch {
            XCTFail("Failed to serialize JSON into data from json: \(validTestJSON)")
        }
    }
    
    func testDecodeFromDataFailure() {
        
        do {
            let data = try JSONSerialization.data(withJSONObject: invalidTestJSON)
            
            let testNil = Test(from: data)
            
            XCTAssertNil(testNil)
        }
        catch {
            XCTFail("Failed to serialize JSON into data from json: \(validTestJSON)")
        }
    }
    
    func testEncode() {
        
        let testMatch = Test(testID: 1, email: "test@email.com", testOptional: "optional", testEnum: TestStringEnum.testMatch)
        
        guard var encodedJson = testMatch.jsonValue else {
            XCTFail("Failed to encode Test Object")
            return
        }
        
        XCTAssertEqual(encodedJson["testID"] as? Int, 1)
        XCTAssertEqual(encodedJson["email"] as? String, "test@email.com")
        XCTAssertEqual(encodedJson["testOptional"] as? String, "optional")
        XCTAssertEqual(encodedJson["testEnum"] as? String, "testMatch")
        
        let testNoMatch = Test(testID: 3, email: "test@test.com", testOptional: "optional", testEnum: TestStringEnum.testNoMatch)
        
        guard var encodedJsonNoMatch = testNoMatch.jsonValue else {
            XCTFail("Failed to encode Test Object")
            return
        }
        
        XCTAssertEqual(encodedJsonNoMatch["testID"] as? Int, 3)
        XCTAssertEqual(encodedJsonNoMatch["email"] as? String, "test@test.com")
        XCTAssertEqual(encodedJson["testOptional"] as? String, "optional")
        XCTAssertEqual(encodedJsonNoMatch["testEnum"] as? String, "NoMatch")
        
        let testNil = Test(testID: 5, email: "test@nil.com", testOptional: nil, testEnum: TestStringEnum.testMatch)
        
        guard var encodedJsonWithNil = testNil.jsonValue else {
            XCTFail("Failed to encode Test Object")
            return
        }
        
        XCTAssertEqual(encodedJsonWithNil["testID"] as? Int, 5)
        XCTAssertEqual(encodedJsonWithNil["email"] as? String, "test@nil.com")
        XCTAssertNil(encodedJsonWithNil["testOptional"])
        XCTAssertEqual(encodedJsonWithNil["testEnum"] as? String, "testMatch")
    }
    
}


MetovaJSONCodableTests/CustomChildProtocolTests.swift
//
//  CustomChildProtocolTests.swift
//  MetovaJSONCodableTests
//
//  Created by Kalan Stowe on 4/12/18.
//  Copyright © 2018 Metova Inc.
//
//  MIT License
//
//  Permission is hereby granted, free of charge, to any person obtaining
//  a copy of this software and associated documentation files (the
//  "Software"), to deal in the Software without restriction, including
//  without limitation the rights to use, copy, modify, merge, publish,
//  distribute, sublicense, and/or sell copies of the Software, and to
//  permit persons to whom the Software is furnished to do so, subject to
//  the following conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
//  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
//  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
//  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import XCTest
@testable import MetovaJSONCodable

struct TestDateWithTimestamp: JSONEpochDateCodable {
    var testID: Int
    var email: String
    var testOptional: String?
    var testDate: Date
}

struct TestDateWithFormatter: JSONCustomDateFormatterCodable {
    var testID: Int
    var email: String
    var testDate: Date
    
    static var dateFormatter: DateFormatter {

        let dateFormatter = DateFormatter()
        dateFormatter.calendar = Calendar(identifier: .iso8601)
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
        return dateFormatter
    }
}

class CustomChildProtocolTests: XCTestCase {
 
    let validTestJSON: JSON = [
        "testID": 1,
        "email": "test@email.com",
        "testOptional": "optional",
        "testDate": 1523560533
    ]
    
    let invalidTestJSON: JSON = [
        "testID": 2,
        "email": "test@email.com",
        "testOptional": "optional",
        "testDate": "2001-09-11T17:56:27.000Z"
    ]
    
    let validTestJSONISO8601: JSON = [
        "testID": 2,
        "email": "test@email.com",
        "testDate": "2001-09-11T17:56:27.000Z"
    ]
    
    let invalidTestJSONISO8601: JSON = [
        "testID": 1,
        "email": "test@email.com",
        "testDate": 1523560533
    ]
    
    func testDecodeTimestampSuccess() {
        
        guard let testMatch = TestDateWithTimestamp(from: validTestJSON) else {
            XCTFail("Failed to decode Test object from json: \(validTestJSON)")
            return
        }
        
        XCTAssertEqual(testMatch.testID, 1)
        XCTAssertEqual(testMatch.email, "test@email.com")
        XCTAssertEqual(testMatch.testOptional, "optional")
        XCTAssertEqual(testMatch.testDate, NSDate(timeIntervalSince1970: 1523560533) as Date)
    }
    
    func testDecodeTimestampFailed() {
        
        let testInvalidDate = TestDateWithTimestamp(from: invalidTestJSON)
        
        XCTAssertNil(testInvalidDate)
    }
    
    func testEncodeTimestampSuccess() {
        
        let date = Date()
        
        let testMatch = TestDateWithTimestamp(testID: 3, email: "test@test.com", testOptional: nil, testDate: date)
        guard var encodedJson = testMatch.jsonValue else {
            XCTFail("Failed to encode Test Object")
            return
        }
        
        XCTAssertEqual(encodedJson["testID"] as? Int, 3)
        XCTAssertEqual(encodedJson["email"] as? String, "test@test.com")
        XCTAssertEqual(encodedJson["testDate"] as? Double, Double(date.timeIntervalSince1970))
    }
    
    func testDecodeWithDateFormatterSuccess() {
        
        guard let testMatch = TestDateWithFormatter(from: validTestJSONISO8601) else {
            XCTFail("Failed to decode Test object from json: \(validTestJSON)")
            return
        }
        
        let compareDate = TestDateWithFormatter.dateFormatter.date(from: "2001-09-11T17:56:27.000Z")
        
        XCTAssertEqual(testMatch.testID, 2)
        XCTAssertEqual(testMatch.email, "test@email.com")
        XCTAssertEqual(testMatch.testDate, compareDate)
    }
    
    func testDecodeWithDateFormatterFailed() {
        
        let testInvalidDate = TestDateWithFormatter(from: invalidTestJSONISO8601)
        
        XCTAssertNil(testInvalidDate)
    }
    
    func testEncodeDateFormatterSuccess() {
        
        let date = Date()
        
        let testMatch = TestDateWithFormatter(testID: 10, email: "new@test.com", testDate: date)
        
        let compareDate = TestDateWithFormatter.dateFormatter.string(from: date)
        
        guard var encodedJson = testMatch.jsonValue else {
            XCTFail("Failed to encode Test Object")
            return
        }
        
        XCTAssertEqual(encodedJson["testID"] as? Int, 10)
        XCTAssertEqual(encodedJson["email"] as? String, "new@test.com")
        XCTAssertEqual(encodedJson["testDate"] as? String, compareDate)
    }
}
